pyomo/contrib/pynumero/interfaces/nlp_transformations.py

```diff
@@ -376,45 +368,330 @@ def hessian_lag(self, x, y, out=None, **kwargs):
+
+        """
+        return empty_matrix(0, self.nx)
+
+    def jacobian_d(self, x, out=None, **kwargs):
+
+        """Return the jacobian of inequality constraints evaluated at x
+
+        Parameters
+        ----------
+        x : 1d-array
+            array with values of primal variables. Size nx
+        out : 1d-array
+            coo_matrix with the structure of the jacobian already defined. Optional
+
+        Returns
+        -------
+        The jacobian of the inequality contraints in a coo_matrix format
+
+        """
+        return self._base_nlp.jacobian_d(x, out=out, **kwargs)
+
+    def hessian_lag(self, x, y, out=None, **kwargs):
+        # y should only include yd
+        assert y.size == self.nd
+        Pc = self.Pc
+        Pd = self.Pd
+        yd = Pd.dot(y)
+        y_tilde = self._w_values + self.rho * self._base_nlp.evaluate_c(x)
+        yc = Pc.dot(y_tilde)
+        yall = yd + yc
+        hess = self._base_nlp.hessian_lag(x, yall, **kwargs)
+        jac = self._base_nlp.jacobian_c(x)
+        rjac = self.rho * jac.transpose().dot(jac)
+        res = hess + rjac
+        return res.tocoo()
+
+    def expansion_matrix_c(self):
+        """
+        Returns expansion matrix inequality constraints
+        """
+        row = np.zeros(0)
+        nnz = 0
+        col = np.arange(nnz, dtype=np.int)
+        data = np.zeros(nnz)
+        return csr_matrix((data, (row, col)), shape=(self.ng, nnz))
+
+    def expansion_matrix_d(self):
+        """
+        Returns expansion matrix inequality constraints
+        """
+        row = self._base_nlp._d_map
+        nnz = len(self._base_nlp._d_map)
+        col = np.arange(nnz, dtype=np.int)
+        data = np.ones(nnz)
+        return csr_matrix((data, (row, col)), shape=(self.ng, nnz))
+
+    def expansion_matrix_du(self):
+        """
+        Returns expansion matrix upper bounds on inequality constraints
+        """
+        row = self._base_nlp._upper_d_map
+        nnz = len(self._base_nlp._upper_d_map)
+        col = np.arange(nnz, dtype=np.int)
+        data = np.ones(nnz)
+        return csr_matrix((data, (row, col)), shape=(self.nd, nnz))
+
+    def expansion_matrix_dl(self):
+        """
+        Returns expansion matrix lower bounds on inequality constraints
+        """
+
+        row = self._base_nlp._lower_d_map
+        nnz = len(self._base_nlp._lower_d_map)
+        col = np.arange(nnz, dtype=np.int)
+        data = np.ones(nnz)
+        return csr_matrix((data, (row, col)), shape=(self.nd, nnz))
+
+    def expansion_matrix_xu(self):
+        """
+        Returns expansion matrix for upper bounds on primal variables
+        """
+        return self._base_nlp.expansion_matrix_xu()
+
+    def expansion_matrix_xl(self):
+        """
+        Returns expansion matrix for lower bounds on primal variables
+        """
+        return self._base_nlp.expansion_matrix_xl()
 
 
+    #jac = nlp.jacobian_c(x)
+    #print(jac.todense())
```

Comment:
Should these be deleted or moved into `if __name__ == "__main__"`?

Reply:
Done